### PR TITLE
Jm auth

### PR DESCRIPTION
**Pull request by jmangel**
First pull request for #12

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -11,4 +11,6 @@
       post :share, to: 'images#send_email'
     end
   end
+
+  resources :users
```
Comment (bboe) [inline config/routes.rb]: This commit does not contain any controllers at this point. Consider adding one controller action at a time in a commit so that your commits are atomic (focus on one thing) and each commit delivers value.

File: app/assets/javascripts/users.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (bboe) [inline app/assets/javascripts/users.js]: Is this file necessary? In general, only keep files around that are actually being utilized.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
+class UsersController < ApplicationController
+  def show
```
Comment (bboe) [inline app/controllers/users_controller.rb]: What's the purpose of the user show page?

File: test/controllers/users_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  test 'invalid signup information' do
+    assert_no_difference 'User.count' do
+      post users_path, params: { user: { name:  '',
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: What's the response status?

File: test/controllers/users_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+require 'test_helper'
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Can you add a controller test for users/new?

File: config/routes.rb (hunk 1 above)
Comment (bboe) [reply config/routes.rb]: Also, will you be using all the routes that this generates? If not, consider adding `only: [...]` to list the ones that you need.

File: test/controllers/users_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  test 'invalid signup information' do
+    assert_no_difference 'User.count' do
+      post users_path, params: { user: { name:  '',
+                                         email: 'user@invalid',
+                                         password:              'foo',
+                                         password_confirmation: 'bar' } }
+    end
+    assert_select 'form[action="/users"]'
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Great assertion! Consider using the path helpers rather than hard-coding the path.

File: test/flows/user_crud_test.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  test 'sign up' do
+    users_new_page = PageObjects::Users::NewPage.visit
+
+    users_new_page = users_new_page.create_user!(
+      name: '',
+      email: '',
+      password: '',
+      password_confirmation: ''
+    ).as_a(PageObjects::Users::NewPage)
+    assert_equal "can't be blank", users_new_page.name.error_message
```
Comment (bboe) [inline test/flows/user_crud_test.rb]: Since you should already be testing these messages via controller tests, there's little reason to test them in the flow-test. In general, the flow test need only test the go-right case.

File: test/models/user_test.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                     password: 'foobar', password_confirmation: 'foobar')
+  end
+
+  test 'should be valid' do
+    assert @user.valid?
+  end
+
+  test 'name should be present' do
+    @user.name = '     '
+    assert_not @user.valid?
```
[the remark this answers is not in the document]
Comment (jmangel) [reply test/models/user_test.rb]: Cool, I'm working on this now. For email address I have two validators: presence and email_format. Email_format (from a gem) also fails when the email is empty. Should I test with `assert_equal({ email: ["can't be blank", "does not appear to be a valid e-mail address"] }, @user.errors.messages)` or should I get rid of the redundant presence validator?

@bboe

File: db/migrate/20171005195712_add_index_to_users_email.rb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+class AddIndexToUsersEmail < ActiveRecord::Migration[5.0]
```
[the remark this answers is not in the document]
Comment (jmangel) [reply db/migrate/20171005195712_add_index_to_users_email.rb]: Oops, must've gotten distracted while doing this one. Fixed

[the remark this answers is not in the document]
Comment (jmangel) [reply db/migrate/20171005195712_add_index_to_users_email.rb]: Oops, must've gotten distracted while doing this one. Fixed

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (jmangel) [reply app/controllers/users_controller.rb]: Currently it just does nothing but display "Success!" I wasn't planning on creating a more detailed view until later in the issue

File: test/controllers/users_controller_test.rb (hunk 4 above)
Comment (jmangel) [reply test/controllers/users_controller_test.rb]: I believe here it should still be 200, even though the user validation/creation itself failed. Do you mean I should add `assert_response :ok` here? @bboe

File: test/models/user_test.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (bboe) [reply test/models/user_test.rb]: I would suggest removing the redundant presence validator.

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (bboe) [reply app/controllers/users_controller.rb]: Having it is fine, just make sure there is a controller test for it.

File: test/controllers/users_controller_test.rb (hunk 4 above)
Comment (bboe) [reply test/controllers/users_controller_test.rb]: If the status code should be 200, then yes, there should definitely be a `assert_response :ok`. However, did create occur ok? If not, maybe consider a different response code. `unprocessible_entity` is often used if the validation fails.

File: test/controllers/users_controller_test.rb (hunk 5 above)
Comment (jmangel) [reply test/controllers/users_controller_test.rb]: done

File: db/migrate/20171005195712_add_index_to_users_email.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (bboe) [reply db/migrate/20171005195712_add_index_to_users_email.rb]: This migration can also be collapsed with the create table one.

File: test/models/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  end
+
+  test 'should be valid' do
+    assert_predicate @user, :valid?
+  end
+
+  test 'name should be present' do
+    @user.name = '     '
+    refute_predicate @user, :valid?
+    assert_equal({ name: ["can't be blank"] }, @user.errors.messages)
```
Comment (bboe) [inline test/models/user_test.rb]: Great!

File: test/models/user_test.rb [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+  test 'email should not be too long' do
+    @user.email = 'a' * 244 + '@example.com'
+    refute_predicate @user, :valid?
+    assert_equal({ email: ['is too long (maximum is 255 characters)',
+                           'does not appear to be a valid e-mail address'] },
+                 @user.errors.messages)
+  end
+
+  test 'email with less than maximum length should be valid' do
```
Comment (bboe) [inline test/models/user_test.rb]: Isn't your go-right case already doing this? Since it's not an explicit boundary condition, it isn't necessary to make an additional test for this case.

File: test/models/user_test.rb [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    @user.email = 'a' * 10 + '@example.com'
+    assert_predicate @user, :valid?
+  end
+
+  test 'email validation should accept valid addresses' do
+    valid_addresses = %w(user@example.com USER@foo.COM A_US-ER@foo.bar.org
+                         first.last@foo.jp alice+bob@baz.cn)
+    valid_addresses.each do |valid_address|
+      @user.email = valid_address
+      assert_predicate @user, :valid?, "#{valid_address.inspect} should be valid"
```
Comment (bboe) [inline test/models/user_test.rb]: No need for the message when using `assert_predicate`.

File: app/helpers/users_helper.rb [hunk 13]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
```
Comment (bboe) [inline app/helpers/users_helper.rb]: Is this used?

File: test/controllers/users_controller_test.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+                                         email: 'user@example.com',
+                                         password:              'password',
+                                         password_confirmation: 'password' } }
+    end
+
+    assert_redirected_to User.last
+  end
+
+  test 'should get new' do
+    get new_user_url
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Make sure to always test the response status, either via `assert_response` or `assert_redirected_to`.

File: test/controllers/users_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+    assert_select "form[action='#{users_path}']"
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Perfect form-level assertion!

File: app/views/application/_nav.html.erb [hunk 16]
```diff
@@ -6,11 +6,19 @@
...
     <ul class="navbar-nav mr-auto">
-      <li class="nav-item active">
+      <li class="nav-item <%= add_active_class(:images, :index) %>">
         <%= link_to images_path, class: 'nav-link' do %>
-            Home <span class="sr-only">(current)</span>
+            Home
+            <%= add_sr_only_current(:images, :index) %>
         <% end %>
       </li>
+      <li class="nav-item <%= add_active_class(:users, :new) %> <%= add_active_class(:users, :create) %>">
```
Comment (bboe) [inline app/views/application/_nav.html.erb]: Do you have any sort of tests that rely on the active status of the navbar? Perhaps you could use an `assert_select` when on the relevant page to check to ensure that it is the only item that is active?

File: test/flows/user_crud_test.rb [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    users_new_page = PageObjects::Users::NewPage.visit
+
+    users_show_page = users_new_page.create_user!(
+      name: 'John doe',
+      email: 'john.doe@example.com',
+      password: 'password',
+      password_confirmation: 'password'
+    )
+
+    assert_equal 'You have successfully signed up', users_show_page.flash_message(:success)
```
Comment (bboe) [inline test/flows/user_crud_test.rb]: Great! Make sure to expand on this later (once login is permitted) so that this flow test doesn't mimic exactly something testable via a controller test.

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (jmangel) [reply app/controllers/users_controller.rb]: Done

File: app/helpers/users_helper.rb (hunk 13 above)
Comment (jmangel) [reply app/helpers/users_helper.rb]: Nope, I think I ended up adding the code to application_helper instead. I'll delete this

File: app/views/application/_nav.html.erb (hunk 16 above)
Comment (jmangel) [reply app/views/application/_nav.html.erb]: Yup, just added that assertion to my controller tests to get images index and get new user page

File: test/models/user_test.rb [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+  test 'name should be present' do
+    @user.name = '     '
+    refute_predicate @user, :valid?
+    assert_equal({ name: ["can't be blank"] }, @user.errors.messages)
+
+    @user.name = ''
+    refute_predicate @user, :valid?
+    assert_equal({ name: ["can't be blank"] }, @user.errors.messages)
+
+    @user.name = nil
```
Comment (bboe) [inline test/models/user_test.rb]: Great! While this is perfectly fine, you can also loop through the three cases.
